src/Microsoft.VisualStudio.Editor.Razor/RazorSyntaxTreePartialParser.cs

```diff
@@ -65,5 +71,61 @@ private PartialParseResultInternal GetPartialParseResult(SourceChange change)
 
             return result;
         }
+
+        // Internal for testing
+        internal static SyntaxTreeNode DeepCopy(SyntaxTreeNode node)
+        {
+            switch (node)
+            {
+                case TagHelperBlock tagHelperBlock:
+                    var tagHelperBlockBuilder = new TagHelperBlockBuilder(tagHelperBlock);
+
+                    tagHelperBlockBuilder.Children.Clear();
+                    for (var i = 0; i < tagHelperBlock.Children.Count; i++)
+                    {
+                        var clonedChild = DeepCopy(tagHelperBlock.Children[i]);
+                        tagHelperBlockBuilder.Children.Add(clonedChild);
+                    }
+
+                    tagHelperBlockBuilder.Attributes.Clear();
+                    for (var i = 0; i < tagHelperBlock.Attributes.Count; i++)
+                    {
+                        var existingAttribute = tagHelperBlock.Attributes[i];
+                        var clonedValue = existingAttribute.Value != null ? DeepCopy(existingAttribute.Value) : null;
+                        tagHelperBlockBuilder.Attributes.Add(
+                            new TagHelperAttributeNode(existingAttribute.Name, clonedValue, existingAttribute.AttributeStructure));
+                    }
+
+                    if (tagHelperBlock.SourceStartTag != null)
+                    {
+                        var clonedStartTag = (Block)DeepCopy(tagHelperBlock.SourceStartTag);
+                        tagHelperBlockBuilder.SourceStartTag = clonedStartTag;
+                    }
+
+                    if (tagHelperBlock.SourceEndTag != null)
+                    {
+                        var clonedEndTag = (Block)DeepCopy(tagHelperBlock.SourceEndTag);
+                        tagHelperBlockBuilder.SourceEndTag = clonedEndTag;
+                    }
+
+                    return tagHelperBlockBuilder.Build();
+                case Block block:
+                    var blockBuilder = new BlockBuilder(block);
+
+                    blockBuilder.Children.Clear();
+                    for (var i = 0; i < block.Children.Count; i++)
+                    {
+                        var clonedChild = DeepCopy(block.Children[i]);
+                        blockBuilder.Children.Add(clonedChild);
+                    }
+
+                    return blockBuilder.Build();
+                case Span span:
+                    var spanBuilder = new SpanBuilder(span);
+                    return spanBuilder.Build();
+                default:
```

Comment:
This functionality should be virtual methods on these classes, not a switch statement.

Reply:
Gonna make it abstract on SyntaxTreeNode